Add support for registering scripted IOOptionsWidget

jcfr opened the pull request.

lassoan commented on Base/QTCore/qSlicerCoreIOManager.cxx:
```diff
@@ -1133,6 +1133,18 @@ void qSlicerCoreIOManager::registerIO(qSlicerIO* io)
 }
 
+//-----------------------------------------------------------------------------
+int qSlicerCoreIOManager::numberOfRegisteredFileReader(const qSlicerIO::IOFileType& fileType) const
```
We should follow Qt conventions in Qt classes, which means that we should use `registeredFileReaderCount` name.
```suggestion
int qSlicerCoreIOManager::registeredFileReaderCount(const qSlicerIO::IOFileType& fileType) const
```

lassoan commented on Base/QTCore/qSlicerCoreIOManager.cxx:
```diff
@@ -1133,6 +1133,18 @@ void qSlicerCoreIOManager::registerIO(qSlicerIO* io)
+}
+
+//-----------------------------------------------------------------------------
+int qSlicerCoreIOManager::numberOfRegisteredFileWriter(const qSlicerIO::IOFileType& fileType) const
```
```suggestion
int qSlicerCoreIOManager::registeredFileWriterCount(const qSlicerIO::IOFileType& fileType) const
```

lassoan commented on Base/QTCore/qSlicerCoreIOManager.h:
```diff
@@ -214,6 +214,17 @@ class Q_SLICER_BASE_QTCORE_EXPORT qSlicerCoreIOManager:public QObject
+  /// \a qSlicerScriptedLoadableModule::registerIO()
+  /// \a qSlicerScriptedLoadableModule::readers()
+  /// \a qSlicerScriptedLoadableModule::writers()
+  Q_INVOKABLE int numberOfRegisteredFileReader(const qSlicerIO::IOFileType& fileType) const;
```
change these to `...count`

lassoan commented on Base/QTCore/qSlicerScriptedFileReader.cxx:
```diff
@@ -239,6 +239,12 @@ QStringList qSlicerScriptedFileReader::extensions()const
+//-----------------------------------------------------------------------------
+qSlicerIOOptions* qSlicerScriptedFileReader::options()const
+{
+  return nullptr;
```
Describe here and/or in the header file that returning nullptr is not an error it simply means that there are no options for this reader.

jcfr replied:
Additional updates will be pushed to effectively implements this. In the mean time, the PR is in draft.

lassoan commented on Base/QTGUI/qSlicerScriptedIOOptionsWidget.cxx:
```diff
@@ -0,0 +1,169 @@
+  Py_DECREF(arguments);
+  if (!result)
+    {
+    // Method call failed (probably an omitted function), call default implementation
```
I know that we do this all the time, but now that you are working on this and you are so knowledgable about Python integration, I would ask: is it possible to do this better? The main limitation of this is that if an uncaught exception is raised in the reimplemented method then it is silently ignored and it is assumed that the method is just not overridden.

jcfr replied:
There is way to address this and simplified across Slicer, will do so in a follow-up PR.

lassoan commented on Base/QTGUI/qSlicerScriptedIOOptionsWidget.h:
```diff
@@ -0,0 +1,69 @@
+  /// \sa qSlicerIO::updateGUI()
+  void updateGUI(const qSlicerIO::IOProperties& ioProperties) override;
+
+  /// Return the ioProperties when exec() is being called.
```
Extra comment or missing implementation.

lassoan reviewed (commented):
Thank you, this looks great, just added a few comments inline.

jcfr committed:
ENH: Relocate qSlicerScriptedFileReader to QTGUI

This change is implemented anticipating the integration of scripted
IOOptionsWidget that will required to override the options() method
and instantiate a qSlicerScriptedIOOptionsWidget object.

jcfr committed:
ENH: Add support for implementing scripted IOOptionsWidget

jamesobutler commented:
I have rebased this PR to help with testing. Note that former commits in this PR were already integrated through PR #6993. Commits listed below:
https://github.com/Slicer/Slicer/commit/18346ffb632177391fcbe24a27d33fbe2cb763e1
https://github.com/Slicer/Slicer/commit/5540c8c1e56792068f07c689d3650124d35cc057
https://github.com/Slicer/Slicer/commit/fafcc749374234a9569567d54e3cd9788a2b90c7
https://github.com/Slicer/Slicer/commit/e386f989fe13df3454eeee9d10196d481e694417
https://github.com/Slicer/Slicer/commit/be954d7b2841bffeac213cce161c8e05c2880f0c
https://github.com/Slicer/Slicer/commit/99eb7808a7bd6482147fd8c0c21d85f487644544

jcfr commented:
@jamesobutler  This is now ready for review & testing.

jamesobutler commented on Applications/SlicerApp/Testing/Python/ScriptedModuleDiscoveryTest/ModuleG_WithFileReaderAndIOOptions_WithoutWidget.py:
```diff
@@ -0,0 +1,52 @@
+        return True
+
+
+class ModuleG_WithFileReaderAndIOOptions_WithoutWidgetIOOptionsWidget:
+    def __init__(self, parent):
+        print("Registered ModuleG_WithFileReaderAndIOOptions_WithoutWidgetFileReaderIOOptionsWidget")
+        self.parent = parent
+
+    def updateGUI(self):
+        pass
```
@jcfr Can you provide an example of how to say add a checkbox to the scripted file reader's options area? I'm not sure what to put here. A UI file can be loaded here of various widgets?

jamesobutler replied:
👋🏻 @jcfr , I can help test to see if this work is sufficient from my Python scripted file reader that I’d like to add some load options widgets.

Can you help guide how to use this work? Hopefully we can get this older work integrated. Thanks!

jcfr replied:
> Can you provide an example of how to say add a checkbox to the scripted file reader's options area? I'm not sure what to put here. A UI file can be loaded here of various widgets?

Good point. I will look into updating the test accordingly

jamesobutler commented:
@jcfr A few merge conflicts here with various recent integrations.

jcfr commented on Base/QTGUI/qSlicerScriptedLoadableModule.cxx:
```diff
@@ -220,10 +221,23 @@ void qSlicerScriptedLoadableModule::registerIO()
     qSlicerApplication::application()->ioManager()->registerIO(fileWriter.take());
     }
   QScopedPointer<qSlicerScriptedFileReader> fileReader(new qSlicerScriptedFileReader(this));
+  qSlicerScriptedFileReader* registeredFileReaderRegistered = nullptr;
```
`registeredFileReaderRegistered` -> `registeredFileReader`

jcfr committed:
ENH: Progress toward adding a working example of scripted IOOptionsWidget

jamesobutler commented on Applications/SlicerApp/Testing/Python/ScriptedModuleDiscoveryTest/ModuleG_WithFileReaderAndIOOptions_WithoutWidget.py:
```diff
@@ -0,0 +1,84 @@
+
+        self.setupUi()
+
+        self.coordinateSystemComboBox.connect("currentIndexChanged(int)", self.updateProperties)
```
To match the name you gave in setupUI():
```suggestion
        self.coordindateSystemlabelComboBox .connect("currentIndexChanged(int)", self.updateProperties)
```

jamesobutler commented on Applications/SlicerApp/Testing/Python/ScriptedModuleDiscoveryTest/ModuleG_WithFileReaderAndIOOptions_WithoutWidget.py:
```diff
@@ -0,0 +1,84 @@
+
+    def setupUi(self):
+        layout = qt.QHBoxLayout()
+        parent.setLayout(layout)
```
```suggestion
        self.parent.setLayout(layout)
```

jamesobutler reviewed (commented):
@jcfr Thanks for moving this work forward 🙏🏻 . Using this example I was able to see the progress you made to have options for a scripted file reader:
![image](https://github.com/Slicer/Slicer/assets/15837524/be10cf88-63f3-4763-83ba-3740aaf1e332)

jcfr committed:
ENH: Progress toward adding a working example of scripted IOOptionsWidget (2)

jamesobutler commented on Applications/SlicerApp/Testing/Python/ScriptedModuleDiscoveryTest/ModuleG_WithFileReaderAndIOOptions_WithoutWidget.py:
```diff
@@ -0,0 +1,79 @@
+        pass
+
+    def updateProperties(self):
+        self.parent.properties["coordinateSystem"] = vtkMRMLStorageNode.GetCoordinateSystemTypeFromString(
```
Trying this update in my custom scripted file reader, it appears that the properties dictionary isn't updated with these new custom entries. Oddly reviewing `self.parent.properties` at the end of this method was revealing that the new entry wasn't added and therefore not used in the FileReader's `load` method which then uses `properties` as an input argument.

jcfr replied:
>  it appears that the properties dictionary isn't updated

Thanks for testing :pray: I also observed this. I have an idea of the issue and will follow up.

jcfr committed:
ENH: Progress toward adding a working example of scripted IOOptionsWidget (3)

Support updating "Properties" by updating qSlicerScriptedIOOptionsWidget
to derive its pimpl from qSlicerIOOptionsPrivate and updating the Q_PROPERTY
declaration to use "setProperties" as a write method.

jcfr commented:
@jamesobutler What we are still missing is the ability to implement `examineFileInfoList` from Python. 

Specialization may allow to:
- preprocess the list of file
- define new properties  (the method `updateGui` is called automatically after `examineFileInfoList` allowing the widget to update itself based on the new properties)

For example:

https://github.com/Slicer/Slicer/blob/516716c035743cd26364a7e7aefb72940d8abc8c/Modules/Loadable/Volumes/qSlicerVolumesReader.cxx#L219-L253
